sw/host/rom_ext_image_tools/signer/src/main.rs

```diff
@@ -9,91 +9,177 @@
-    let device_usage_value =
-        &device_usage_value(&config.input_files.usage_constraints_path);
-    let system_state_value =
-        &system_state_value(&config.input_files.system_state_value_path);
+use anyhow::bail;
+use anyhow::ensure;
+use anyhow::Context;
+use anyhow::Result;
+
+// Type aliases for convenience.
+type ImageSignature =
+    mundane::public::rsa::RsaSignature<B3072, RsaPkcs1v15, Sha256>;
+type PrivateKey = mundane::public::rsa::RsaPrivKey<B3072>;
+
+// TODO: Remove this struct when this functionality is integrated into `opentitantool`.
+struct Args {
+    input: PathBuf,
+    priv_key: PathBuf,
+    output: PathBuf,
+}
 
-    let mut message_to_sign = Vec::<u8>::new();
-    message_to_sign.extend_from_slice(system_state_value);
-    message_to_sign.extend_from_slice(device_usage_value);
-    message_to_sign.extend_from_slice(image_sign_data);
+impl Args {
+    pub fn new(args: env::ArgsOs) -> Result<Args> {
+        let args = args.skip(1).collect::<Vec<_>>();
+        match args.as_slice() {
+            [input, priv_key, output] => Ok(Args {
+                    input: PathBuf::from(input),
+                    priv_key: PathBuf::from(priv_key),
+                    output: PathBuf::from(output),
+                }),
+            args => bail!("Expected exactly 3 positional arguments: input, private key, and output, got: {}.", args.len()),
+        }
+    }
+}
 
-    let signature = RsaSignature::<B3072, RsaPkcs1v15, Sha256>::sign(
-        &private_key,
-        &message_to_sign,
-    )
-    .expect("Failed to sign!");
+/// Parses an unsigned big-endian hex string into a little-endian byte vector.
+fn parse_hex_str(hex_str: &str) -> Result<Vec<u8>> {
+    ensure!(
+        hex_str.starts_with("0x")
+            && hex_str.len() > 2
+            && hex_str.len() % 2 == 0,
+        "Invalid hex string: {}",
+        hex_str
+    );
+    let bytes = (2..hex_str.len())
+        .step_by(2)
+        .map(|i| u8::from_str_radix(&hex_str[i..i + 2], 16))
+        .rev()
+        .collect::<Result<_, _>>()?;
+    Ok(bytes)
+}
 
-    image.update_signature_field(&signature.bytes());
+/// Updates the manifest of an image.
+// TODO: This function must use a public key.
+fn update_image_manifest(
+    image: &mut Image,
+    key: impl AsRef<Path>,
+) -> Result<()> {
+    let key = fs::read(&key).with_context(|| {
+        format!("Failed to read the key from `{}`.", key.as_ref().display())
+    })?;
+    let key =
+        PrivateKey::parse_from_der(&key).context("Failed to parse the key.")?;
+
+    image.set_manifest_field(
+        &manifest::ROM_EXT_SIGNATURE_KEY_PUBLIC_EXPONENT,
+        key.public_exponent_be().iter().rev().cloned(),
+    )?;
+    image.set_manifest_field(
+        &manifest::ROM_EXT_SIGNATURE_KEY_MODULUS,
+        key.public_modulus_be().iter().rev().cloned(),
+    )?;
+    image.set_manifest_field(
+        &manifest::ROM_EXT_MANIFEST_IDENTIFIER,
+        parse_hex_str("0x4552544f")?,
```

Comment:
nit: The native rust integer types have `to_xx_bytes` methods:

https://doc.rust-lang.org/std/primitive.u32.html#method.to_le_bytes

For fields which are "rust integer sized", we can avoid string parsing.  Whether this is valuable depends on how some of these fields would interact with user-supplied arguments.

Reply:
Generally, my opinion is that the CLI interface to signing in `opentitantool` should be concerned with parsing string arguments into whatever is the ideal form for the signing code.

Just  to harp on the `timestamp` field: I don't want to enter that in hexadecimal.  The `date` command emits an integer in decimal representation and if I choose to use a CL number in place of the date field, that is also emitted by the tools as decimal.  The CLI parser should be concerned with examining user and looking for numerical-base identifiers like `0x`, etc.

Reply:
Replaced all `parse_hex_str` calls with `to_le_bytes` except for this one (just to call `parse_hex_str` once in this file). We can decide if we want to keep/enhance/remove this function when we start integrating the signer into opentitantool.